pkg/pdp/service/roots_add.go

```diff
@@ -132,41 +154,70 @@ func (p *PDPService) AddRoots(ctx context.Context, id uint64, request []types.Ro
 
 		// Generate the unsealed CID from the collected piece infos.
 		proofType := abi.RegisteredSealProof_StackedDrg64GiBV1_1
-		generatedCID, err := nonffi.GenerateUnsealedCID(proofType, pieceInfos)
+		generatedCID, err := GenerateUnsealedCID(proofType, pieceInfos)
+		if err != nil {
+			return common.Hash{}, fmt.Errorf("failed to generate RootCID: %w", err)
+		}
+
+		// turn the uploaded roots into PieceCIDV1
+		providedPieceCidV1, err := asPieceCIDv1(addReq.Root.String())
 		if err != nil {
-			return common.Hash{}, fmt.Errorf("failed to generate RootCID: %v", err)
+			return common.Hash{}, fmt.Errorf("failed to generate PieceCIDV1 for request: %w", err)
 		}
+
 		// Compare the generated and provided CIDs.
-		if !addReq.Root.Equals(generatedCID) {
-			return common.Hash{}, fmt.Errorf("provided RootCID does not match generated RootCID: %s != %s", addReq.Root, generatedCID)
+		if !providedPieceCidV1.Equals(generatedCID) {
+			return common.Hash{}, fmt.Errorf("provided RootCID does not match generated RootCID: %s (v1 %s) != %s",
+				addReq.Root, providedPieceCidV1, generatedCID)
 		}
 	}
 
 	// Step 5: Prepare the Ethereum transaction data outside the DB transaction
 	// Obtain the ABI of the PDPVerifier contract
-	abiData, err := contract.PDPVerifierMetaData()
+	abiData, err := smartcontracts.PDPVerifierMetaData()
 	if err != nil {
 		return common.Hash{}, fmt.Errorf("failed to get abi data from PDPVerifierMetaData: %w", err)
 	}
 
-	// Prepare RootData array for Ethereum transaction
-	// Define a Struct that matches the Solidity RootData struct
-	type RootData struct {
-		Root struct {
-			Data []byte
-		}
-		RawSize *big.Int
+	// Prepare PieceData array for Ethereum transaction
+	// Define a Struct that matches the Solidity PieceData struct
+	type PieceData struct {
+		Data []byte // CID
 	}
 
-	var rootDataArray []RootData
+	var pieceDataArray []PieceData
 
 	for _, addRootReq := range request {
 		// Convert RootCID to bytes
 		rootCID := addRootReq.Root
 
+		_, rawSize, err := commcid.PieceCidV1FromV2(rootCID)
+		if err != nil {
+			return common.Hash{}, fmt.Errorf("invalid PieceCIDV2: %w", err)
+		}
+		height, _, err := commcid.PayloadSizeToV1TreeHeightAndPadding(rawSize)
+		if err != nil {
+			return common.Hash{}, fmt.Errorf("computing height and padding: %w", err)
+		}
+		if height > 50 {
```

Comment:
What is magic 50?

Reply:
Sorry, forgot to add a comment about this.

it's defined here: 
https://github.com/FilOzone/pdp/blob/main/src/PDPVerifier.sol#L46

and checked here:
https://github.com/FilOzone/pdp/blob/main/src/PDPVerifier.sol#L492